site/app/controllers/calendar/CalendarController.php

```diff
@@ -27,8 +31,125 @@ class CalendarController extends AbstractController {
+        $type = $_POST['type'];
+        $date = $_POST['date'];
+        $text = $_POST['text'];
+
+        if (strip_tags($text) !== $text) {
+            $this->core->addErrorMessage("HTML cannot be used in this text");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        $calendar_item = new CalendarItem();
+        try {
+            $calendar_item->setStringType($type);
+        }
+        catch (\InvalidArgumentException $e) {
+            $this->core->addErrorMessage("That is not a valid calendar item type");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+        $calendar_item->setDate(new \DateTime($date));
+        try {
+            $calendar_item->setText($text);
+        }
+        catch (\InvalidArgumentException $e) {
+            $this->core->addErrorMessage("Text exceeds 255 characters which is not allowed");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        $this->core->getCourseEntityManager()->persist($calendar_item);
+        $this->core->getCourseEntityManager()->flush();
+
+        $this->core->addSuccessMessage("Calendar item successfully added");
+        return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/calendar/editItem", methods={"POST"})
+     * @AccessControl(role="INSTRUCTOR")
+     */
+    public function editMessage(): RedirectResponse {
+        $type = $_POST['type'];
+        $date = $_POST['date'];
+        $text = $_POST['text'];
+        $id = $_POST['id'];
+
+        if (strip_tags($text) !== $text) {
+            $this->core->addErrorMessage("HTML cannot be used in this text");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        $calendar_item = $this->core->getCourseEntityManager()->getRepository(CalendarItem::class)
+            ->findOneBy(['id' => $id]);
+
+        if ($calendar_item === null) {
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        $calendar_item->setText($text);
+        $calendar_item->setDate(new \DateTime($date));
+        try {
+            $calendar_item->setStringType($type);
+        }
+        catch (\InvalidArgumentException  $e) {
+            $this->core->addErrorMessage("That is not a valid calendar item type");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+        $this->core->getCourseEntityManager()->flush();
+        return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/calendar/deleteItem", methods={"POST"})
+     * @AccessControl(role="INSTRUCTOR")
+     */
+    public function deleteMessage(): JsonResponse {
+        $id = $_POST['id'];
```

Comment:
Same comment about checking for the existence of the field before blindly accessing it here.